pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,405 @@
+	nnLowWatermark = 2 // the number of peers in consecutive deepest bins that constitute as nearest neighbours
+)
+
+var (
+	errMissingAddressBookEntry = errors.New("addressbook underlay entry not found")
+	timeToRetry                = 30 * time.Second
+)
+
+type binSaturationFunc func(bin, depth uint8, peers *pslice.PSlice) bool
+
+// Options for injecting services to Kademlia.
+type Options struct {
+	Base           swarm.Address
+	Discovery      discovery.Driver
+	AddressBook    addressbook.Interface
+	P2P            p2p.Service
+	SaturationFunc binSaturationFunc
+	Logger         logging.Logger
+}
+
+// Kad is the Swarm forwarding kademlia implementation.
+type Kad struct {
+	base           swarm.Address         // this node's overlay address
+	discovery      discovery.Driver      // the discovery driver
+	addressBook    addressbook.Interface // address book to get underlays
+	p2p            p2p.Service           // p2p service to connect to nodes with
+	saturationFunc binSaturationFunc     // pluggable saturation function
+	connectedPeers *pslice.PSlice        // a slice of peers sorted and indexed by po, indexes kept in `bins`
+	knownPeers     *pslice.PSlice        // both are po aware slice of addresses
+	depth          uint8                 // current neighborhood depth
+	depthMu        sync.RWMutex          // protect depth changes
+	manageC        chan struct{}         // trigger the manage forever loop to connect to new peers
+	waitNext       map[string]time.Time  // sancation connections to a peer, key is overlay string and value is time to next retry
+	waitNextMu     sync.Mutex            // synchronize map
+	logger         logging.Logger        // logger
+	quit           chan struct{}         // quit channel
+	done           chan struct{}         // signal that `manage` has quit
+}
+
+// New returns a new Kademlia.
+func New(o Options) *Kad {
+	if o.SaturationFunc == nil {
+		o.SaturationFunc = binSaturated
+	}
+
+	k := &Kad{
+		base:           o.Base,
+		discovery:      o.Discovery,
+		addressBook:    o.AddressBook,
+		p2p:            o.P2P,
+		saturationFunc: o.SaturationFunc,
+		connectedPeers: pslice.New(maxBins),
+		knownPeers:     pslice.New(maxBins),
+		manageC:        make(chan struct{}, 1),
+		waitNext:       make(map[string]time.Time),
+		logger:         o.Logger,
+		quit:           make(chan struct{}),
+		done:           make(chan struct{}),
+	}
+
+	go k.manage()
+	return k
+}
+
+// manage is a forever loop that manages the connection to new peers
+// once they get added or once others leave.
+func (k *Kad) manage() {
+	var peerToRemove swarm.Address
+	defer close(k.done)
+
+	for {
+		select {
+		case <-k.quit:
+			return
+		case <-k.manageC:
+			err := k.knownPeers.EachBinRev(func(peer swarm.Address, po uint8) (bool, bool, error) {
+				if k.connectedPeers.Exists(peer) {
+					return false, false, nil
+				}
+
+				k.waitNextMu.Lock()
+				if next, ok := k.waitNext[peer.String()]; ok && time.Now().Before(next) {
```

Comment:
yeah but within the slice the peers are not ordered according to anything expect PO, i.e. within the po there is no ordering

Reply:
The earlier implementation had a struct for each peer instead of just peer inside a bin. Each struct would also contain the connection status (which you maintain globally). 

We can still do it here by iterating waitNext and picking peers which were never tried.